### PR TITLE
[FEAT][Performance]Optimize native sampling by removing expensive scatter

### LINKED ISSUE #27710: [Feature]: Faster apply_top_k_top_p without scatter
### 🚀 The feature, motivation and pitch

In `apply_top_k_top_p`, the following code unpermutes the sorted logits:
`logits = logits_sort.scatter(dim=-1, index=logits_idx, src=logits_sort)`.
Then we randomly sample a token from the distribution by
`return probs.div(q).argmax(dim=-1).view(-1)`.

The two steps above are equivalent to 
`idx_permuted = probs.div(q).argmax(dim=-1).view(-1)`
`return logits_idx[idx_permuted]` which (almostly) saves a `scatter` operation. `idx_permuted` is much smaller than `logits_sort`.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Flink-ddd**
## Purpose

This PR optimizes the native PyTorch sampling implementation (specifically for Top-P/Nucleus sampling) by replacing the expensive scatter operation with a lightweight gather approach.

**Motivation:**
Currently, apply_top_k_top_p sorts the logits for filtering and then uses scatter to restore the probabilities to the original vocabulary size (Batch, Vocab). This scatter operation involves large, non-contiguous memory writes, which becomes a performance bottleneck, especially for models with large vocabulary sizes (e.g., Qwen2.5 with ~152k vocab).

**Proposed Change:**
Instead of scattering back to the full vocabulary size before sampling:
1. We perform sampling directly on the sorted probabilities (reduced search space).
2. We use the resulting sampled indices (which are relative to the sorted array) to **gather** the actual token IDs from the sorted_indices.
3. This transforms an $O(B \times V)$ write operation into an $O(B)$ read operation.

**Safety & Compatibility:**
* **Top-K only:** The logic falls back to the original path (no sorting), ensuring zero regression for pure Top-K sampling.
* **Logprobs/CPU:** Added a `scatter_if_sorted` flag to ensure backward compatibility for cases requiring full logprob distributions or CPU execution.

Closes #27710

## Test Plan

I benchmarked this change on an **NVIDIA RTX A6000 (48GB)** using **Qwen/Qwen2.5-7B-Instruct** (Vocab size: 152k) to highlight the impact on large vocabularies.

**Methodology:**
1. **Performance:** Used vllm bench serve with dataset-name random to measure Throughput and TPOT across three scenarios: Top-P (Target), Top-K (Regression Check), and Top-P+K.
2. **Correctness:** Used lm_eval on the gsm8k dataset to ensure generation quality remains consistent.

## Test Result

**1. Performance (vLLM Serving Benchmark)**
Observed a consistent throughput improvement (~1.9% - 2.0%) in scenarios involving sorting (Top-P). Pure Top-K performance remains unaffected (no regression).

### Benchmark Serving (NVIDIA RTX A6000     model: Qwen2.5-7B-Instruct)

Mode | Metric | Before | After | Speed-up (×) | Change
-- | -- | -- | -- | -- | --
top-p | Output Throughput (tok/s) | 2079.99 | 2118.59 | 1.02× | up 1.86%
  | Total Throughput (tok/s) | 4159.98 | 4237.19 | 1.02× | up 1.86%
  | TTFT (ms) | 3654.56 | 3612.18 | 1.01× | up 1.16%
  | TPOT (ms) | 113.75 | 111.69 | 1.02× | up 1.81%
  | ITL (ms) | 113.75 | 111.69 | 1.02× | up 1.81%
top-k | Output Throughput (tok/s) | 2349.74 | 2345.00 | 1.00× | down 0.20%
  | Total Throughput (tok/s) | 4699.48 | 4691.00 | 1.00× | down 0.18%
  | TTFT (ms) | 3605.74 | 3588.81 | 1.00× | up 0.47%
  | TPOT (ms) | 100.10 | 100.54 | 1.00× | down 0.44%
  | ITL (ms) | 100.10 | 100.55 | 1.00× | down 0.45%
top-p+k | Output Throughput (tok/s) | 2070.59 | 2112.93 | 1.02× | up 2.05%
  | Total Throughput (tok/s) | 4141.18 | 4223.87 | 1.02× | up 2.00%
  | TTFT (ms) | 3629.26 | 3595.21 | 1.01× | up 0.94%
  | TPOT (ms) | 114.40 | 112.20 | 1.02× | up 1.92%
  | ITL (ms) | 114.40 | 112.20 | 1.02× | up 1.92%

</body></html>

**2. Correctness (GSM8K via lm_eval)**
Verified that accuracy remains consistent.
Note: Since torch.multinomial results depend on element order, sampling from sorted logits vs. original logits may yield different tokens for the same seed, but the statistical distribution remains identical.

### LM_EVAL (NVIDIA RTX A6000   gsm8k   model: Qwen2.5-7B-Instruct)

Mode | Metric | Before | After | Change
-- | -- | -- | -- | --
top-p | Exact Match (%) | 0.8271 | 0.8287 | up 0.19%
  | Std Err | 0.0104 | 0.0104 | 0.00%
  | Strict Match | 0.7665 | 0.7665 | 0.00%
  | Std Err | 0.0117 | 0.0117 | 0.00%
top-k | Exact Match (%) | 0.8271 | 0.8264 | down 0.08%
  | Std Err | 0.0104 | 0.0104 | 0.00%
  | Strict Match | 0.7640 | 0.7627 | down 0.17%
  | Std Err | 0.0117 | 0.0117 | 0.00%
top-p+k | Exact Match (%) | 0.8271 | 0.8287 | up 0.19%
  | Std Err | 0.0104 | 0.0104 | 0.00%
  | Strict Match | 0.7650 | 0.7672 | up 0.29%
  | Std Err | 0.0117 | 0.0116 | down 0.85%

</body></html>

### CODE REVIEW

**Comment by Flink-ddd** [comment]
@codex review

**Comment by Flink-ddd** [comment]
Hi @njhill , Could pls you review this pr When you have free time? Thanks.
